### ISSUE TITLE
Running packaged version as non-privileged user

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#638]
Right now, instructions on running LeoFS as non-privileged user are flaky at best: it's not directly mentioned anywhere, there is mention of "chmod" in instructions describing installation of package (which, obviously, is quite random and files should come with correct ownership directly from package), but without details it's more or less worthless - it's not obvious how to run scripts correctly under non-root user and so on. From my experiments, absolutely nothing prevents any kind of node to run as user, it's just that the way default packages are created that makes it complicated.

This is serious problem when running cluster on supported Linux versions (Ubuntu, Debian and RHEL-based); it obviously doesn't exist when developing / running from source. According to sources, it's also not a problem in FreeBSD port where maintainer provided patches for lots of files which solves both problems of running as non-privileged and upgrades. I'm willing to help with fixing packages and/or scripts to solve it for Linux as well; is there interest in doing this? Because, obviously, some things will have to be changed in documentation and default options and I will need some help and advices from other developers / users.

Also, these changes are related to ones that can make version upgrade easier (right now, upgrade process for production clusters, as per documentation is very complicated and error-prone; but automating it isn't that easy as well because of few issues complicating this). However, change of upgrade process is more complicated issue, I don't think it should be touched at least for now.

There are few ways to accomplish comfortable usage of all LeoFS nodes as non-priveleged user under Linux. I see at least 3 ways (depending the way cluster is managed and SA preference):

1. **Running with data/queue/log files in-tree, that is, /usr/local/leofs/<version> hierarchy.**

_Required changes_: all packaged-installed leofs hierarchy (or part of it, e.g. excluding bin/, erts-*/ and lib/) is owned by "leofs" user. Requires few changes to packaging, package scripts to add user. Requires setting of RUNNER_USER in launch scripts and some fixes related to that (more on this later).

_Advantages_: simple, minimal changes needed, great for experiments and evaluating.

_Disadvantages_: for real production, and planned version upgrades nearly anyone would want to keep dynamically changed data/queue/log directories out of package-installed tree which changes every version, so this design is mostly useless for production systems.

2. **Running with etc/data/queue/log and the rest of dynamically changed directories out of tree, but using default launch scripts.**

_Required changes_: current launch scripts don't support passing of environment options to them or reading them from any config file. There is no way to set RUNNER_USER from outside the script, and there is no way to set RUNNER_LOG_DIR - defaults value of which wants to create in-tree erlang.log and run_erl.log files, no way to set RUNNER_ETC_DIR in which vm.args and app.config are created. Stopping (but not starting!) of node causes error "Failed to create cookie file '/usr/local/leofs/.erlang.cookie': eaccesescript: exception error: no match of right hand side value" if /usr/local/leofs is not user-writable - I'm not sure how to fix this, e.g. by forcing to create this "cookie" not in the tree but in /tmp or /var/run. Plus, general problem of using RUNNER_USER in script has to be fixed.

_Advantages_: package-shipped binaries and out-of-tree config and data files is most likely way current packages will be used in production. It's possible to do this with changes only to launch script, leaving everything else to SA (creation of user, out-of-tree directories with appropriate permissions and so on). Not that it *should* be done this way, but it's possible, at least, without breaking anything.

_Disadvantages_: all directories in default config file needs to be changed, environment variables passed to script and so on - fine for production, but too complicated for easy evaluation / testing. There are few unobvious and annoying moments like requirement to write to "etc" directory because app.config and vm.args are rewritten on launch, changes to snmp_manager config are required for keeping "db" directory out of tree and such.

3. **Same as 2., but using customized launch script**

Some system administrators might decide to use their own copy of launch scripts, at fixed path, to have full customization over directories, have less problems during package changes and other reasons. Now, I'm not 100% sure it's desirable to pursue this way.

_Required changes_: the way RUNNER_SCRIPT_DIR is calculated right now doesn't work when script isn't placed in "bin" directory, this can be fixed, e.g. by giving ability to pass it through environment variable or config file. Scripts are different for different types of nodes: manager node support 3 additional commands (no big deal, it's possible to have this in merged script), launch command is different (different schema and config files) - again, it's possible to add ability to redefine that with environment variable.

There is also this difference which I don't understand:
```
$ diff -U0 leo_storage/bin/leo_storage leo_gateway/bin/leo_gateway
...
@@ -215 +215 @@
-        CMD="$BINDIR/erlexec -boot $RUNNER_BASE_DIR/releases/$APP_VSN/$BOOTFILE -mode minimal -config $CONFIG_PATH -args_file $VMARGS_PATH -- ${1+"$@"}"
+        CMD="$BINDIR/erlexec -boot $RUNNER_BASE_DIR/releases/$APP_VSN/$BOOTFILE -mode embedded -config $CONFIG_PATH -args_file $VMARGS_PATH -- ${1+"$@"}"
```

I wonder if it's possible to run gateway in the same mode as others. Not that this difference is impossible to work around, but, well, I'm just wondering if it's really needed and should be left as is.

_Advantages_: more agility. Potentially, same script for all means less code duplication and less possible bugs coming from having almost the same code 3 times, so everyone benefits from this change. Of course, bash isn't exactly the nicest language and there is a chance that resulting version will be slightly more ugly, but still.

_Disadvantages_: like I said before, I'm not perfectly sure that someone will be actually using this (we planned to, but decided to stick to way "2" at least for now. That might change, however). Changing something without existing use case isn't a good idea. Maybe this change (merging of scripts) can cause some problems that I don't see right now.

If doing these changes, there is no need to stick to just one - 1) and 2) can be both done, and maybe 3) as well. There are also general problems with launcher scripts, I have noticed at least two. First, RUNNER_USER is utilized through "exec sudo -i" command, which forces RUNNER_USER to have working shell (it can't be /sbin/nologin, which is usually used). This isn't ground-breaking, but would be nice if fixed. Second, launching script with relative path won't work if RUNNER_USER is defined (since it starts a new shell and passes relative path to it). Pretty easy to fix.

Anyhow, please share your thoughts on these changes. I might be missing some use-cases and important points here as well, would be glad to hear if developers or other users notice anything. If there is interest in merging these kind of changes if they are done properly, I'll start working on that.

**Comment by mocchira** [comment]
@vstax Thanks for you suggestion.

> is there interest in doing this? Because, obviously, some things will have to be changed in documentation and default options and I will need some help and advices from other developers / users.

Yes at least I have. (and believe lots here

> If doing these changes, there is no need to stick to just one - 1) and 2) can be both done

+1 with implementing both 1, 2 and keep the default 1 to ensure the backward compatibility and allow users to opt-in 2 as needed.
(maybe adopting 3 In future after taking deep considerations and experiments

The below are replies to your questions.

> Stopping (but not starting!) of node causes error "Failed to create cookie file '/usr/local/leofs/.erlang.cookie': eaccesescript: exception error: no match of right hand side value" if /usr/local/leofs is not user-writable - I'm not sure how to fix this, e.g. by forcing to create this "cookie" not in the tree but in /tmp or /var/run.
  
This should be fixed when RUNNER_USER is set to an appropriate user as .erlang.cookie file is created at `/${HOME}/.erlang.cookie`.

> There are few unobvious and annoying moments like requirement to write to "etc" directory because app.config and vm.args are rewritten on launch, changes to snmp_manager config are required for keeping "db" directory out of tree and such.

Yes definitely annoying to SA.
However it's needed to provide SA with SA-friendly ini-like file format in .conf files instead of Erlang Native file format like app.config that is difficult to edit properly.
it's kind of hack that makes life a little easier for SA but unfortunately also can be annoying for SA like you when they tried to hack around here ;)

> There is also this difference which I don't understand
> I wonder if it's possible to run gateway in the same mode as others. Not that this difference is impossible to work around, but, well, I'm just wondering if it's really needed and should be left as is.

IIRC,
https://github.com/leo-project/leofs/commit/561c314fab1371acf20b8690fc0ea581ce2e8ba3
As the above commit implied,
Without minimal options,
As leo_storage can receive erlang RPC requests from other nodes in a cluster before initializing inner state, that cause leo_storage to have inconsistent state and may result in crash so yes this is still needed.

> There are also general problems with launcher scripts, I have noticed at least two. First, RUNNER_USER is utilized through "exec sudo -i" command, which forces RUNNER_USER to have working shell (it can't be /sbin/nologin, which is usually used). This isn't ground-breaking, but would be nice if fixed.  Second, launching script with relative path won't work if RUNNER_USER is defined (since it starts a new shell and passes relative path to it). Pretty easy to fix.

Totally agreed to be fixed accordingly.

**Comment by vstax** [comment]
Thanks for explanation. It doesn't seem like LeoFS uses .erlang.cookie for anything other than "ping" to local process (I mean as a concept as documented at http://erlang.org/doc/getting_started/conc_prog.html - there is no need to have same cookie on various nodes and so on). It's just created with random string on the first run and that's it. This (somewhat) interferes with 2), since normally there is no need to write to this file other than its first creation, so I'll generate it by filling with some random string during package installation - if it doesn't exist yet. The way I see it, it shouldn't affect existing users, whether they have this file or not.

**Comment by vstax** [comment]
@mocchira 
Both 1) and 2) work fine on my systems (rpm-based, but nothing should break on Debian, I'll do a quick testing for that later), I'm still doing more testing before I will post the code, let me outline a few things to make sure I'm moving into right direction.

### 1) - running in current hierarchy but as leofs user.
User/group "leofs" are created during package installation with home at /usr/local/leofs. File .erlang.cookie is created in there with random data (just some random md5, this works for its purpose), unless it exists, then it's just checked for correct ownership / permissions and they are changed if needed (so if someone was already using a user with home in that directory, nothing would break).

Package spec becomes a bit more complicated since now it needs different permissions for some parts of hierarchy, that is, leo_*/etc, leo_*/log, leo_*/log/sasl, leo_*/snmp/snmpa_*/db, leo_*/work/* are now owned by "leofs" user. For storage node, "avs" directory is created, owned by leofs as well. There is no need to change permissions for /usr/local/leofs itself (since once cookie exists, it's only read).

Startup scripts are modified to run as "leofs" user (can be changed in separate, new "environment" config file). This change requires mention in documentation / release notes, because it changes the upgrade process. I'll write a more detailed explanation, but basically existing users running leofs as root have 3 options (note that new users don't need to do anything - after installation, everything will just run as leofs user):

1. If they were running with files in-tree, during the upgrade - after they copy/move existing "work", "avs", "log", "etc" directories to new tree - they have to change owner of these directories to "leofs" after copy
1. If they were running with files outside of installation tree and don't need to copy them, they should change owner of these files to "leofs" user
1. If they don't want to change the way anything runs, they should set RUNNER_USER to "root" in environment file and it will work the same

The last (most compatible) option is not a default because running as root even once will break owner on various files and one won't be able to run as leofs user anymore without changing it all manually, so this will only postpone the migration and won't allow new users to just run it as non-root.

If existing users were running current packages as non-root with the help of some wrapping scripts or anything, they should set RUNNER_USER to that user (unless it's called "leofs" already) in environment file, and, in theory, that's it. Wrapper script to change user (like the one FreeBSD package uses) can be removed or simplified after that.

### 2) - running in out-of-tree hierarchy to simplify upgrades

In each "etc" directory, there is now a file for re-defining environment variables:
```
$ cat /usr/local/leofs/1.3.2.1/leo_manager_1/etc/leo_manager.environment 
RUNNER_LOG_DIR=/var/log/leofs/leo_manager_1
RUNNER_ETC_DIR=/etc/leofs/leo_manager_1
# This is default
# RUNNER_USER=leofs

[root@leo-m1 1.3.2.1]# ll /var/log/leofs/leo_manager_1
итого 72
drwxr-xr-x. 2 leofs leofs 24576 мар 18 12:00 app
-rw-r--r--. 1 leofs leofs 27796 мар 18 12:45 erlang.log.1
drwxr-xr-x. 2 leofs leofs  4096 мар 17 18:51 ring
-rw-r--r--. 1 leofs leofs   301 мар 17 20:51 run_erl.log
drwxr-xr-x. 2 leofs leofs  4096 мар 17 20:51 sasl
[root@leo-m1 1.3.2.1]# ll /etc/leos/leo_manager_1
итого 16
-rw-r--r--. 1 leofs leofs 1607 мар 17 20:51 app.config
-rw-r--r--. 1 leofs leofs 5770 мар 17 18:53 leo_manager.conf
-rw-r--r--. 1 leofs leofs  289 мар 17 20:51 vm.args
```
In there, it's possible to redefine RUNNER_USER, including setting it back to "root", and set the directories of "erlang.log" / "run_erl.log", also set directory for config files (now separated from schema in etc/ directory - the option to moving it exists as well, if needed). The rest of directories (mnesia, queues, app / sasl logs, crash dump and ring dump directories) can already be set through config file.

I will post a short guide for documentation with example of how to set all the options to run completely out-of-tree as well. For this type of installation, "leo_*.environment" file is the only one SA needs to change after installing the new version during upgrade, the rest is picked through it. It would be really nice if global environment file worked as well, but I can't think of a clean way to do it while keeping the current ability to run both management nodes, multiple gateways or multiple storage nodes on the same system (or even all of it on the single system). So it's separate environment files in the "etc" directory for each node, at least for now.

### 3)

While not thinking on how to run scripts from different directory  for now (it has a set of its own problems), the fact that there are so many scripts to change is painful (well, not changing itself, but the commit will have the same change 9 times?). I need to do a few changes to the launcher script. E.g. there are pkg/leo_gateway/deploy/sbin/leo_gateway, rel/leo_gateway/files/leo_gateway and apps/leo_gateway/rel/files/leo_gateway - that's 9 in total, if counting for all nodes! I mean, I can just change the one that's used for Linux packages, but will that really be enough to cover all the cases?

Anyhow, I was thinking on how to unify the code without breaking anything else, at least. One way is to have the same script for all management/gateway/storage nodes, and few minor changes between them (extra options for management node, and different launch command with different config file names) can be picked from script name: kind of like vim/view/ex, bzip/bunzip2 and so on. An extra argument to force it through command line and not script name can be added so that it's possible to use these scripts with totally random name.

**Comment by mocchira** [comment]
@vstax 

> 1) - running in current hierarchy but as leofs user.

All specs looks good to me. (More benefits than drawbacks breaking backward compatibility in some cases)
Please go forward with this direction.

> 2) - running in out-of-tree hierarchy to simplify upgrades

Almost LGTM.

> It would be really nice if global environment file worked as well, but I can't think of a clean way to do it while keeping the current ability to run both management nodes, multiple gateways or multiple storage nodes on the same system (or even all of it on the single system). So it's separate environment files in the "etc" directory for each node, at least for now.

I also can't so yes please go with this spec for now.
(maybe come up with something during the review process.

> 3) get the code less duplicated

Agreed the concept.
Please give us a little time to validate the idea carefully.

**Comment by vstax** [comment]
@mocchira 
For instance, the fact that "shift" is commented here (only in leo_manager script! It's not the case for leo_gateway and leo_storage, despite exactly the same command afterwards) really looks like example of a bug coming from the fact that scripts are separate (version in "apps" has no history so I link to this one):
https://github.com/leo-project/leofs/blame/master/pkg/leo_manager/deploy/sbin/leo_manager#L104

Regarding unifying scripts, here is how I see it:
1) Versions in "pkg/leo_*/deploy" seem to be obsolete? It looks like SmartOS-specific version but I was unable to verify if it's used right now or it picks the same scripts as other systems right now. I'm not sure where actual build scripts for SmartOS are (I've found https://github.com/project-fifo/mi-leofs/tree/master/copy/root - according to which it should use the same scripts as everything else, I think, but not sure this is actual version). Either way, it might be best not to touch these scripts at all and leave the job of merging them - if needed - to SmartOS maintainer
2) Versions in "rel" seem to be generated by copying them from "apps". I don't quite understand why "rel/leo_*/files" is under version control system as well, if it's replaced by files from apps on each build? Shouldn't they be removed?
3) Here is example of single script that suits all types of nodes - copy/link it to leo_manager, leo_gateway and leo_storage and it will work for each case without anyone noticing a change.. This one is not 100% perfect ("Usage" text in the end of script is incomplete, also it allows dump-mnesia-data etc command on non-manager nodes) but otherwise works on any kind of node. Additionally, it supports specifying type of node with command-line argument, if someone wants to use it without renaming.
```
--- ./apps/leo_manager/rel/files/leo_manager	2017-01-24 23:26:59.356157723 +0300
+++ leo_manager	2017-03-21 21:41:56.564795956 +0300
@@ -24,6 +24,44 @@
 # Identify the script name
 SCRIPT=`basename $0`
 
+case "$1" in
+    -type)
+        case "$2" in
+            leo_manager|leo_gateway|leo_storage)
+                NODE_TYPE=$2
+                shift 2
+                ;;
+            *)
+                echo "Usage: $SCRIPT [-type leo_manager|leo_gateway|leo_storage] {start|stop|ping|remote_console}"
+                echo "Script type is picked from its name, but can be overriden with -type option"
+                exit 1
+                ;;
+        esac
+        ;;
+    *)
+        case "$SCRIPT" in
+            leo_manager|leo_gateway|leo_storage)
+                NODE_TYPE=$SCRIPT
+                ;;
+            *)
+                echo "Script name (currently: $SCRIPT) should be one of: leo_manager, leo_gateway, leo_manager"
+                echo "Or '-type leo_manger|leo_gateway|leo_storage' has to be supplied as the first option"
+                exit 1
+                ;;
+        esac
+        ;;
+esac
+
+case "$NODE_TYPE" in
+    leo_storage)
+        ERLEXEC_MODE=minimal
+        ;;
+    leo_gateway|leo_manager)
+        ERLEXEC_MODE=embedded
+        ;;
+esac
+
+
 # Parse out release and erts info
 START_ERL=`cat $RUNNER_BASE_DIR/releases/start_erl.data`
 ERTS_VSN=${START_ERL% *}
@@ -53,7 +91,7 @@
     rm -f $RUNNER_BASE_DIR/etc/app.*.config
     rm -f $RUNNER_BASE_DIR/etc/vm.*.args
     ERTS_PATH=$RUNNER_BASE_DIR/erts-$ERTS_VSN/bin
-    RES_CUTTLEFISH=`PATH=$ERTS_PATH:$PATH $RUNNER_BASE_DIR/bin/cuttlefish -i $RUNNER_BASE_DIR/etc/leo_manager.schema -c $RUNNER_BASE_DIR/etc/leo_manager.conf -d $RUNNER_BASE_DIR/etc/`
+    RES_CUTTLEFISH=`PATH=$ERTS_PATH:$PATH $RUNNER_BASE_DIR/bin/cuttlefish -i $RUNNER_BASE_DIR/etc/$NODE_TYPE.schema -c $RUNNER_BASE_DIR/etc/$NODE_TYPE.conf -d $RUNNER_BASE_DIR/etc/`
 
     APP_CONFIG=`find $RUNNER_BASE_DIR/etc/ -type f | grep app.*[0-9].config`
     mv $APP_CONFIG $RUNNER_BASE_DIR/etc/app.config
@@ -246,7 +284,7 @@
         BINDIR=$ROOTDIR/erts-$ERTS_VSN/bin
         EMU=beam
         PROGNAME=`echo $0 | sed 's/.*\\///'`
-        CMD="$BINDIR/erlexec -boot $RUNNER_BASE_DIR/releases/$APP_VSN/$BOOTFILE -mode embedded -config $CONFIG_PATH -args_file $VMARGS_PATH -- ${1+"$@"}"
+        CMD="$BINDIR/erlexec -boot $RUNNER_BASE_DIR/releases/$APP_VSN/$BOOTFILE -mode $ERLEXEC_MODE -config $CONFIG_PATH -args_file $VMARGS_PATH -- ${1+"$@"}"
         export EMU
         export ROOTDIR
         export BINDIR
```

**Comment by mocchira** [comment]
@vstax 

> For instance, the fact that "shift" is commented here (only in leo_manager script! It's not the case for leo_gateway and leo_storage, despite exactly the same command afterwards) really looks like example of a bug coming from the fact that scripts are separate (version in "apps" has no history so I link to this one):

Absolutely.

> Versions in "pkg/leo_*/deploy" seem to be obsolete? It looks like SmartOS-specific version but I was unable to verify if it's used right now or it picks the same scripts as other systems right now. I'm not sure where actual build scripts for SmartOS are (I've found https://github.com/project-fifo/mi-leofs/tree/master/copy/root - according to which it should use the same scripts as everything else, I think, but not sure this is actual version). Either way, it might be best not to touch these scripts at all and leave the job of merging them - if needed - to SmartOS maintainer

Maybe obsolete and yes it's SmartOS-specific version so 

> Either way, it might be best not to touch these scripts at all and leave the job of merging them - if needed - to SmartOS maintainer

Please go with this direction.

> Versions in "rel" seem to be generated by copying them from "apps". I don't quite understand why "rel/leo_*/files" is under version control system as well, if it's replaced by files from apps on each build? Shouldn't they be removed?

The files under "rel/leo_\*/files" is the master and have been used to generate Erlang release files.
The files under "apps/leo_\*/rel/files" have existed due to the historical reason that leo_(manager|storage|gateway) each had been managed in its own repository.  Now we have all components under the right one place so we can make those files managed by each component obsolete (to be removed) and make files under "rel/leo_*/files" being the only files to be maintained.

> Here is example of single script that suits all types of nodes - copy/link it to leo_manager, leo_gateway and leo_storage and it will work for each case without anyone noticing a change.. This one is not 100% perfect ("Usage" text in the end of script is incomplete, also it allows dump-mnesia-data etc command on non-manager nodes) but otherwise works on any kind of node. Additionally, it supports specifying type of node with command-line argument, if someone wants to use it without renaming.

Looks good to me.
However this change also require erlang specific release scripts named "reltool.config" such as https://github.com/leo-project/leofs/blob/1.3.2.1/rel/leo_manager/reltool.config to be changed and now we are considering to get those obsolete and adopt the new one managed by https://github.com/erlware/relx so to avoid confusing, it might be good no to try deduplicating at this timing.

**Comment by vstax** [comment]
@mocchira Thanks for explanation. By looking at Makefile and reltool config I was thinking that version from "apps" replaces version from "rel" but now I see that it's not the case. Still, config files are picked from under "apps"...
I need to change reltool.config anyway so it's not a big deal, I think scripts can be unified before switching to new release script, but sure, let's do it as a next step. I'll be doing the same changes for all 3 scripts from "rel" directory for now.

For now I have posted implementation of 1) and 2) to https://github.com/vstax/leofs/tree/develop and https://github.com/vstax/leofs_package/tree/experimental - please try it out when you can and share your thoughts.
Problems of current implementation:
1) rpm complains about defining "snmp/*/db" directories twice. I think it can be left as is for now, this warning can be removed by listing all files directly which will add even more mess to the .spec file
2) just rpm version for now; need to make sure it works before touching Debian package
3) the .environment file is commited in 3 copies, just like scripts; but it's the same script. It should be the first candidate for deduplication - just copy it 3 times during release with different names, but I'm not sure where to place it and where to perform this deduplication - in Makefile or reltool.config? Can you please advise?

Regarding the order, changes https://github.com/vstax/leofs/commit/cda2c0aaef0683122a1591edaf6a2455193c54a4 and https://github.com/vstax/leofs/commit/2592934d32f55caceb7fab2b31efecfd3437027e should be merged before any package changes and are supposed not to break anything and be independent of each other.

Package change https://github.com/vstax/leofs_package/commit/8e05195a2b4c6c87474f93fbe75308994d98b758 (and other package changes, like Debian) should go after, they require script changes to work, and should be merged right before https://github.com/vstax/leofs/commit/d90aed17fb2ac8dca33f88395a160c04adb6b9b3 change (without which packages won't work correctly.. probably).

**Comment by mocchira** [comment]
@vstax Thank you!
Almost LGTM.

> Thanks for explanation. By looking at Makefile and reltool config I was thinking that version from "apps" replaces version from "rel" but now I see that it's not the case. Still, config files are picked from under "apps"...

Yes, the current structure inclines many people to be confused so dedupe as much would be important task and ...

> I need to change reltool.config anyway so it's not a big deal, I think scripts can be unified before switching to new release script, but sure, let's do it as a next step. I'll be doing the same changes for all 3 scripts from "rel" directory for now.

I've checked your code out and build/release/run that has given me confidence to do dedupe in this phase so let's dedupe now!

> rpm complains about defining "snmp/*/db" directories twice. I think it can be left as is for now, this warning can be removed by listing all files directly which will add even more mess to the .spec file

Reproduced on my dev box.
leaving as is would look good to me.

> the .environment file is commited in 3 copies, just like scripts; but it's the same file. It should be the first candidate for deduplication - just copy it 3 times during release with different names, but I'm not sure where to place it and where to perform this deduplication - in Makefile or reltool.config? Can you please advise?

So we have to define where to place files that are used commonly across any component in making release files.
To me, `${LEOFS_REPO_HOME}/rel/common` looks good however to be honest I'm not familiar with how release files should be organized in Erlang so that I'd like to ask @yosukehara and @windkit to hear their opinion for this problem.
Please drop your comment > both

> Regarding the order, changes vstax@cda2c0a and vstax@2592934 should be merged before any package changes and are supposed not to break anything and be independent of each other.
> Package change vstax/leofs_package@8e05195 (and other package changes, like Debian) should go after, they require script changes to work, and should be merged right before vstax@d90aed1 change (without which packages won't work correctly.. probably).

Got it.

Two feedback there.

1. Now the owner of `%{target_dir}/leo_*/etc/leo_*.(conf|schema)` files is `root:root` (the parent directory's one is `leofs:leofs` though), 
would it be good for those also belonging to `leofs:leofs` for editing in future? 

2. There is a hidden (not well known) requirement for LeoFS that it should be able to build/run on OSX platform only for the evaluation purpose. We have to be explicit it somewhere in the document anyway, what I want to say is we have to avoid using some commands that don't exist on vanilla OSX. In this case, `realpath` is the not portable one. so some hack is needed such as http://stackoverflow.com/questions/1055671/how-can-i-get-the-behavior-of-gnus-readlink-f-on-a-mac 

I think we are now on the right track so please PR, I'd like to review further on PR.

**Comment by vstax** [comment]
@mocchira I've created different set of patches that does deduplication of launch scripts and reading environment files, as well as bug fixes for launch script, using "rel/common" as base directory. They can be moved if this turns out not the best place, but personally I think that this version is quite clean. The PR https://github.com/leo-project/leofs/issues/655 contains first part (which is supposed not to break anything), without the change for `RUNNER_USER` to be `leofs` by default. So for testing RPM package you need to uncomment `RUNNER_USER=leofs` in .environment files of installed packages. Please review; I don't think creating PR for the package scripts makes sense until this part is merged.

Regarding your questions, the idea behind config files owned by root is that software running under non-privileged user isn't supposed to alter its own config, as it can lead to security problems. It might not mean much in this case but is a common practice; the only exception is when software contains code that changes its own config files, e.g. here vm.args is generated so `etc/` directory itself is writable by `leofs` user. It doesn't cause any problems when editing since no one is supposed to edit this file as `leofs` user (system user with `/sbin/nologin` shell). But it's only a matter of preference, if you think that owning whole `etc/` as `leofs:leofs` is more consistent, it can be done this way.

I see about OSX; I've only checked the changes against Solaris and FreeBSD and they both contain `realpath` implementation, I've wrongly assumed that OSX contains it as well if FreeBSD does, sorry about that. I've removed that part, dereferencing symlinks isn't really needed there so I don't need any implementation. Current one shouldn't break except for very exotic cases like executing by feeding the script contents over ssh, which isn't supported in the original anyway.

**Comment by mocchira** [comment]
@vstax 
Thanks for the PR!
We are reviewing now and will feedback if there is something to be fixed.

>  I don't think creating PR for the package scripts makes sense until this part is merged.

Make sense.

> Regarding your questions, the idea behind config files owned by root is that software running under non-privileged user isn't supposed to alter its own config, as it can lead to security problems. It might not mean much in this case but is a common practice; the only exception is when software contains code that changes its own config files, e.g. here vm.args is generated so etc/ directory itself is writable by leofs user. It doesn't cause any problems when editing since no one is supposed to edit this file as leofs user (system user with /sbin/nologin shell). But it's only a matter of preference, if you think that owning whole etc/ as leofs:leofs is more consistent, it can be done this way.

Right.
The reason why I asked this changes is 

> the only exception is when software contains code that changes its own config files

This one.
There were an attempt to change its own config files to reflect the value changed through `leofs-adm update-consistency-level` on the fly.
However as it is opaque whether we will implement this feature or not for the moment,
It seems to be good to leave the owner of *.(conf|schema) as is and changing when it's needed.
So keep owners root:root as your suggestion.

> I've removed that part, dereferencing symlinks isn't really needed there so I don't need any implementation. Current one shouldn't break except for very exotic cases like executing by feeding the script contents over ssh, which isn't supported in the original anyway.

Got it.

**Comment by vstax** [comment]
@mocchira Thanks for the review.

I have created Debian version of the package at https://github.com/vstax/leofs_package
However, I don't really use Debian and have no experience in packaging it, I've read some documentation and did it like that - it seems to work now (the original make script is somewhat unusual in how it doesn't rely on debhelper much but I decided to use it for this change). Do you think it's fine this way?

**Comment by mocchira** [comment]
@vstax 

> it seems to work now (the original make script is somewhat unusual in how it doesn't rely on debhelper much but I decided to use it for this change). Do you think it's fine this way?

Very Fine :)

One thing I noticed while trying your patch is
a post script can fail if leofs user already exists.

```bash
(Reading database ... 241467 files and directories currently installed.)
Preparing to unpack leofs_1.3.2.1-1_amd64.deb ...
Unpacking leofs (1.3.2.1-1) over (1.3.2-1) ...
Setting up leofs (1.3.2.1-1) ...
adduser: The user `leofs' already exists. Exiting.
dpkg: error processing package leofs (--install):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
 leofs
```

Please PR based on the current commit!

**Comment by vstax** [comment]
@mocchira You are right, I've overlooked that. It looked like "adduser" wrapper handles it - and it mostly does, but there is a case when existing user was manually created as non-system (with UID >= 1000), then adding system user with the same name will fail.

PR for package changes is https://github.com/leo-project/leofs_package/pull/4 - but it requires changes to launch.sh that change RUNNER_USER to leofs by default. Need to make sure this PR and that change aren't merged before 1.3.3 release which is in RC2 stage already, though?

**Comment by mocchira** [comment]
@vstax 

> PR for package changes is leo-project/leofs_package#4 - but it requires changes to launch.sh that change RUNNER_USER to leofs by default. Need to make sure this PR and that change aren't merged before 1.3.3 release which is in RC2 stage already, though?

How about checking the version given from the command line argument is
- if -ge 1.3.3 (excluding releasse candidates)
  - build the package with leofs user enabled according to your current PR
- or else
  - display warning messages to notify that
    - leofs user/group will be created for running LeoFS with non-privileged leofs user
    - encourage users to use a prev leofs_package if they want the prev behavior
  - give them the option to choose yes/no to keep going forward building a package as there might be someone who want < 1.3.3 with leofs user enabled.

> I have remembered about one more problem with switching to non-privileged user: if node was launched from root from the same directory before, then changing RUNNER_USER is not enough, even if file permissions for installed directories are correct. Next launch will fail like

Indeed.

> rm -rf /tmp/usr solves it, it's not hard, just not obvious because this error can't be seen in erlang log file, only in syslog. (I wrote it here more as a note to myself to mention it in migration documentation. I don't know of a way to solve this, as removing that directory during package installation is not safe)

As you mentioned, documenting would be the most promising to me.

> Technically, this directory ($PIPE_DIR) is created by launch script, however one can't just remove it at that point (during start) since script re-executes itself and has no permissions to do so, if it was previously created by root. And removing it before re-executing will break that pipe as well. I suppose some trickery is possible around it but not sure it's worth it...

Just documenting would be enough to me.

**Comment by mocchira** [comment]
Note:

> just not obvious because this error can't be seen in erlang log file, only in syslog.

This kind of annoying logging could happen when something went wrong before the Erlang runtime STARTED (reached at some point to get to be able to output erlang log files.
In this case, Failed to create a named pipe for redirecting stdio/stdout to erlang log files.
It would be good practice for any product written in erlang to pay attention not only erlang log files but also syslog if something goes bad.

**Comment by vstax** [comment]
@mocchira "-ge 1.3.3", are you sure about that? Well, I thought it's a bit too late to do these changes since it's already in "RC2" stage. Switching to non-privileged user breaks simple upgrade for certain users (well, you'd expect anyone using complex clustered systems to read Release Notes and follow them, but still), i.e. requires yet-to-be-written documentation changes, plus maybe a bit more time to notice some problems (like: I noticed how I missed "cache" directory of leo_gateway and fixed it only in the last commit). So I expected it to go only into next release - unlike changes related to script deduplication and support for .environment files, which should be safe to go into release since it's (supposedly) does.

But, well, I'm fine with anything, if there is enough time before 1.3.3 release and people responsible for release are OK with these changes.

Regarding version check in script. This is a good idea! I didn't even think of that, forgetting that it's not just .spec file but is used through shell script. But regarding doing it in shell... quite not obvious. I've found this: http://stackoverflow.com/questions/4023830/how-compare-two-strings-in-dot-separated-version-format-in-bash - I like "sort -V" solution there, it should work (exclude RC, like you said). But it's not even clear if this change will go into certain version. Anyhow, if it does - this code works
```
#!/bin/sh

ver_to_compare=$1
supported_version=1.3.3

# Treat 1.3.3-rc as 1.3.2
[ "a$ver_to_compare" != "a${ver_to_compare/1.3.3-rc/}" ] && ver_to_compare=1.3.2

straight=$(printf ${supported_version}\\n${ver_to_compare})
sorted=$(printf ${supported_version}\\n${ver_to_compare} | sort -V)

[ a"$straight" != a"$sorted" ] && echo version too old
```

I've thought of this instead of complicated comparisons and binding to exact version:
`curl -s https://raw.githubusercontent.com/leo-project/leofs/$version/rel/common/launch.sh|grep RUNNER_USER=|grep leofs` - if it returns anything, then it's fine to build package; if not, it's not. Right now it won't allow to build package even for "develop" branch, since RUNNER_USER is not set to "leofs" by default yet. This is imperfect as well since it'll break if the script is moved or anything.

**Comment by mocchira** [comment]
@vstax 

> "-ge 1.3.3", are you sure about that? Well, I thought it's a bit too late to do these changes since it's already in "RC2" stage. Switching to non-privileged user breaks simple upgrade for certain users (well, you'd expect anyone using complex clustered systems to read Release Notes and follow them, but still), i.e. requires yet-to-be-written documentation changes, plus maybe a bit more time to notice some problems (like: I noticed how I missed "cache" directory of leo_gateway and fixed it only in the last commit). So I expected it to go only into next release - unlike changes related to script deduplication and support for .environment files, which should be safe to go into release since it's (supposedly) does.
> But, well, I'm fine with anything, if there is enough time before 1.3.3 release and people responsible for release are OK with these changes.

Now we are going to release 1.3.3 at 4/20/2017 so we have around 2W to work.
It may or may not be sufficient to get the remained tasks done so I'd like to propose that we will ship if we could finish
- [ ] Documentation for LeoFS with non-privilege user
  - [ ] Packaging
  - [ ] Configuration
  - [ ] Migration
  - [ ] FAQ (workarounds like the named pipe case you mentioned)
- [ ] Tests on as many usecases/environments as possible
  - [ ] rpm
  - [ ] deb
 
>I've thought of this instead of complicated comparisons and binding to exact version:
>curl -s https://raw.githubusercontent.com/leo-project/leofs/$version/rel/common/launch.sh|grep RUNNER_USER=|grep leofs - if it returns anything, then it's fine to build package; if not, it's not. Right now it won't allow to build package even for "develop" branch, since RUNNER_USER is not set to "leofs" by default yet. This is imperfect as well since it'll break if the script is moved or anything.

This way looks promising as it can handle even RC cases.

**Comment by vstax** [comment]
@mocchira Sorry for the delay, verifying various things took more time than I expected. Unfortunately, I still need your help about some issues here..

Here is documentation for the package changes. I might need your help to, well, actually insert it in appropriate place. "%version" will have to be changed once it's decided which version we're actually talking about. Also, link to documentation showing show to use .environment files should be inserted there (I can do these things myself once I know where actually this documentation will reside): https://gist.github.com/vstax/38bbf594f3623b84ecd038af315d0568

Here is documentation for .environment files and examples on how to set up .environment and .config files for simple upgrades: https://gist.github.com/vstax/a52f67dcb88ab6e1ae356fb3d5655ccc

Your improvements and fixes are welcome; feel free to change it if something isn't written clear enough (I tried to be as clear as possible, but I'm not a native speaker and am making some mistakes, plus as a result text got a bit too long - I'd love it to be more short and precise but showcasing all possible options made it longer than I wanted it to be).

As for testing, I have performed these experiments and verified that they work correctly when following instructions, no extra steps were needed:
1. Few LeoFS nodes, each running in different VM on Scientific Linux 7.3. Upgrade version 1.3.2.1 running as root to develop version (faked as 1.3.3) which runs as `leofs`; most of working directories out of tree
2. All LeoFS nodes on a single (same SL7.3) system; fresh installation of 1.3.2.1 - install, run, upload/download file. Upgrade to fake 1.3.3, make it run as `leofs` user, make sure can download that file
3. All LeoFS nodes on a single Debian 8 system. Same experiment - upload/download file for freshly installed 1.3.2.1, upgrade to fake 1.3.3, follow the upgrade instructions to make it run as `leofs` user. Make sure all common things work
4. Debian 8 system from previous example, upgrade 1.3.2.1->1.3.3 but make it run as root by setting RUNNER_USER=root. Make sure nothing else is needed. Then make it run as `leofs` user (this one requires more steps, indicated in the end of that document)

What I haven't checked: various complicated leofs-adm and launch scripts commands... because I'm not sure what can be related / broken by this change. I'm sure that the system from experiment №1 is completely fine after the change regarding various status, compaction-related and bucket-related commands and SNMP monitoring works fine as well.

Regarding version check from package scripts. I'm a bit unsure how to actually do it, interface-wise.. I have uploaded the script which works correctly (and has some safety checks so it shouldn't be broken easily). But how to, well, interface with it for desired effect?
``
$ ./check-version.sh 1.3.2
Unable to download file from https://raw.githubusercontent.com/leo-project/leofs/1.3.2/rel/common/launch.sh
Version check is impossible!
$ echo $?
1
$ ./check-version.sh develop
This version is too old to build this package with!
$ echo $?
2
```
It will show that version from https://github.com/vstax/leofs/tree/develop is fine for building new package.

Anyhow, I have created the pull request for original repo: https://github.com/leo-project/leofs/pull/697

**Comment by mocchira** [comment]
@vstax Thank your very much for your great documentation.

> Your improvements and fixes are welcome; feel free to change it if something isn't written clear enough (I tried to be as clear as possible, but I'm not a native speaker and am making some mistakes, plus as a result text got a bit too long - I'd love it to be more short and precise but showcasing all possible options made it longer than I wanted it to be).

Looks great to me.
As I mentioned on each gist,
When I port your documents into our new doc, I may change a little about the sentence structure for the purpose of well suited to our new doc.

As for the English,
I'm also not a native and IMHO your English is much better than me :)
Yes the length of a sentence tend to be long more than necessary however I believe almost everyone can  understand what you explained here.

> What I haven't checked: various complicated leofs-adm and launch scripts commands... because I'm not sure what can be related / broken by this change. I'm sure that the system from experiment №1 is completely fine after the change regarding various status, compaction-related and bucket-related commands and SNMP monitoring works fine as well.

OK.
We will test any other features where you haven't covered. 

> Regarding version check from package scripts. I'm a bit unsure how to actually do it, interface-wise.. I have uploaded the script which works correctly (and has some safety checks so it shouldn't be broken easily). But how to, well, interface with it for desired effect?

Confirmed the check_version.sh works as expected.

As for the **how to actually do it** with interface-wise,
I'm now considering the below

- Tag the current leofs_package as 1.0
- Tag the leofs_package (including your PR) as 1.1
- Include calling check_version.sh into make_{deb|rpm}.sh
- Check the version argument
  - Failed due to
    - curl/head: notify the user there maybe network problem or something goes wrong on your host and recommend to check those.
    - version check: notify the user leofs_package >= 1.1 requires leofs >= 1.3.3 and recommend them to use leofs_package 1.0 if you want to build leofs < 1.3.3
  - Succeeded: go forward making a package

Thoughts?

Also one thing I noticed while checking check_version.sh,
Using trap EXIT improves the code maintainability when you have to do something (`rm -f $tmpfile` in this case) in case exiting a program.

Reference: http://stackoverflow.com/questions/2129923/bash-run-command-before-a-script-exits

**Comment by vstax** [comment]
@mocchira Yes, you are right about "trap" of course. Thanks for suggestion. I have thrown up few checks fast without looking at code later so missed this. Now I've improved script a bit and added its actual execution to scripts. I moved all logic to that script so that only one line is added to the main build scripts. Seems to work for me.

I'm not sure if reporting about network problems will be of any help - it's impossible to check for everything; in fact, even check for curl presence is quite an extra one (one can't remove curl on EL6/EL7 without breaking deps, probably the only case I can think about is really minimal installation of Debian, e.g. in chroot with the purpose to build this package only). The build will fail later during "git clone" anyway.

Your idea about tag & version seems nice; I'm not quite sure it is worth the trouble because, well, is there a demand for building older packaged versions at all? If you think there is, then sure, I have changed the script a bit to adopt to this. Replacing "echo" in go_on_after_warning() - by adding URL to that older version with 1.0 tag or something - should be enough once tags are there, I think.

**Comment by mocchira** [comment]
> I'm not sure if reporting about network problems will be of any help - it's impossible to check for everything; in fact, even check for curl presence is quite an extra one (one can't remove curl on EL6/EL7 without breaking deps, probably the only case I can think about is really minimal installation of Debian, e.g. in chroot with the purpose to build this package only). The build will fail later during "git clone" anyway.

I checked the latest PR and the current error handling looks reasonable to me.

> Your idea about tag & version seems nice; I'm not quite sure it is worth the trouble because, well, is there a demand for building older packaged versions at all? If you think there is, then sure, I have changed the script a bit to adopt to this. Replacing "echo" in go_on_after_warning() - by adding URL to that older version with 1.0 tag or something - should be enough once tags are there, I think.

Tagged the tip on develop at https://github.com/leo-project/leofs_package/tree/1.0.0 so please replace with it.

It seems we finally reach the consensus about spec so I'd like to move on the PR :)

**Comment by vstax** [comment]
@mocchira 
Done

**Comment by mocchira** [comment]
@vstax Thanks! we are very close to the goal :)

Anyway I ported your docs into our new docs with minor modifications (outlines to fit into our new style and little bit sentences to get its length as short as possible).

- https://github.com/leo-project/leofs/blob/develop/docs/admin/system_admin/migration.md 

Please check it out and give me comments if you find something wrong.

P.S.
These docs structured with markdown will be published using http://www.mkdocs.org/ on the our official web site after 1.3.3 release.

**Comment by vstax** [comment]
@mocchira Very nice! Thanks for editing. "Too long sentences" is one of the faults of my writing that I'm familiar with (believe it or not, when editing that documentation before posting it I already broke quite a few sentences into shorter ones..).

I've noticed a few minor mistakes, fixes are at https://github.com/leo-project/leofs/pull/707

However, I think that two things should be improved in this documentation.

First is splitting "Run LeoFS As a Non-Privileged User" and "Environment Config Files" to different pages; right now it's easy to overlook the second part. Migration instructions are only needed for users of versions < 1.3.3 who are upgrading, and they might only need them once; users who are installing 1.3.3 or higher don't need information from that page at all. However, description of "environment config files" might be interesting to users of 1.3.3 and higher as well, so it doesn't quite belong in "migration" documentation (even if there are references to "environment config files" from the "migration" part).

Second is somewhat different header for the first part ("Run LeoFS As a Non-Privileged User"), kind of like "Note" after the diagram on http://leo-project.net/leofs/docs/admin_guide/admin_guide_10.html page; a header that clearly states that this whole section is intended for users of version < 1.3.3 who are upgrading to version >= 1.3.3. This would make it clearer for them that they need to read it, and clear for users of 1.3.3 or higher that they probably don't need to waste their time reading these instructions. I mean, one can understand it while reading but it should be obvious by just looking at header that this is just instruction for migration between specific versions.

What do you think?

**Comment by mocchira** [comment]
> I've noticed a few minor mistakes, fixes are at #707

Thanks! ( merged.

> First is splitting "Run LeoFS As a Non-Privileged User" and "Environment Config Files" to different pages; right now it's easy to overlook the second part. Migration instructions are only needed for users of versions < 1.3.3 who are upgrading, and they might only need them once; users who are installing 1.3.3 or higher don't need information from that page at all. However, description of "environment config files" might be interesting to users of 1.3.3 and higher as well, so it doesn't quite belong in "migration" documentation (even if there are references to "environment config files" from the "migration" part).

Agreed.
Now we are writing the other part of new docs in parallel so we will put the section about environment config files into an appropriate place when the whole doc structure is fixed.

> Second is somewhat different header for the first part ("Run LeoFS As a Non-Privileged User"), kind of like "Note" after the diagram on http://leo-project.net/leofs/docs/admin_guide/admin_guide_10.html page; a header that clearly states that this whole section is intended for users of version < 1.3.3 who are upgrading to version >= 1.3.3. This would make it clearer for them that they need to read it, and clear for users of 1.3.3 or higher that they probably don't need to waste their time reading these instructions. I mean, one can understand it while reading but it should be obvious by just looking at header that this is just instruction for migration between specific versions.

Agreed.
I will revise the header after the other part of this section is fixed.

Thanks for FB.

**Comment by mocchira** [comment]
@windkit @yosukehara 

### TODO
- [ ] @yosukehara 
  - [ ] Confirm make release/bootstrap.sh/mdcr.sh with https://github.com/leo-project/leofs/pull/697 on your local dev env works as expected
  - [ ] build packages on the package build environment and test those packages on appropriate ones

- [ ] @windkit  
  - [ ] Confirm make release/bootstrap.sh/mdcr.sh with https://github.com/leo-project/leofs/pull/697 on your local dev env works as expected
  - [ ] Confirm whether there is no affected parts in CI pipeline by #638 and fix if there is something affected.

- [ ] @mocchira
  - [x] Confirm make release/bootstrap.sh/mdcr.sh with https://github.com/leo-project/leofs/pull/697 on your local dev env works as expected
  - [ ] Revise docs
    - [ ] Separate the Section for Environment Config
    - [ ] Revise the header
    - [ ] Add the caution for developers who have to tweak https://github.com/leo-project/leofs/blob/master/rel/common/launch.environment#L10-L11 while developing with make release/bootstrap.sh/mdcr.sh

**Comment by windkit** [comment]
@mocchira During my test, i have one problem with the `$PIPE_DIR`.

Previously I ran LeoFS with my own account, so the own of `$PIPE_DIR`
After I rebuild with #697 and the deploy, start script would complete without errors but Erlang VM cannot start.
Problem is solved after I remove the `$PIPE_DIR` and let the script recreate it with proper permission.

This should also be added as a notice to developers/users.

**Comment by mocchira** [comment]
@windkit thanks for confirming.

That pitfall already described on https://github.com/leo-project/leofs/blob/develop/docs/admin/system_admin/migration.md#running-leofs-with-default-paths (around remove temporary files) so the current doc have a notice to users however as you pointed out, would be good to have a notice for developers as well.
I will update the doc while taking your opinion into account.

**Comment by vstax** [comment]
@windkit why not just put
```
RUNNER_USER=<myself>
```
in your .bash_profile to keep it working as it was on your development systems? Default settings are for production, but current version of scripts honors any environment settings.

**Comment by windkit** [comment]
@vstax Actually I reverted the change, use leofs as the ansible user fits easily with my scripts.

But it is a good idea to set it in .bash_profile :)

**Comment by mocchira** [comment]
Done [TODO](https://github.com/leo-project/leofs/issues/638#issuecomment-294661617) with https://github.com/leo-project/leofs/commit/1edbc55db0ae6d30688a8d3be416d2e90dee75bc

**Comment by mocchira** [comment]
@vstax 

Since 1.3.3 have landed with non-privilege user,
Please close this issue if you don't have any problem with 1.3.3.

**Comment by vstax** [comment]
Done. Thanks for support!
